### ISSUE TITLE
Ownership of virtual_dom_rs::patch params

### ISSUE THREAD

**Issue by dbrgn** [chinedufn/percy#49]
The `virtual_dom_rs::patch` function takes the `root_node` element by value, not by mutable reference. This means that the element cannot be re-used after patching it.

Is this a conscious design decision? What would be the reason for it?

**Comment by chinedufn** [comment]
It's not mandatory, IIRC I had a lifetime issue and just did it this way to make it work.

I really don't remember the exact issue stumbling block that I ran into back then but it definitely is not desired and I was probably just making some silly mistake.. so if you want to take a stab at making it mutable I'd be happy to answer any questions along the way!

**Comment by dbrgn** [comment]
Will probably give this a try next week!

**Comment by chinedufn** [comment]
Sweet! Again, I really don’t remember the problem, it might not have even been lifetimes and instead have been multiple mutable references or something.

Regardless I’m sure it was just poor form on my end and not a fundamental issue.

Although does this suggest that we might want a reference instead of a miracle reference ...? 

https://github.com/rustwasm/wasm-bindgen/issues/1061

**Comment by dbrgn** [comment]
Yes, a regular reference will probably suffice. The good thing is that the compiler will tell us whether it's needed or not :slightly_smiling_face:

**Comment by dbrgn** [comment]
Hm, it's not that easy, because you are inserting the element into the `nodes_to_patch` hashmap.

This should be doable with only references by rewriting the patching algorithm. I'll try to take a quick look when I find time, but don't know yet if that'll be today or not.

**Comment by dbrgn** [comment]
Ok, changing it to references using the same recursive algorithm is a bit tricky because of the lifetimes involved.

Since the `Element` is clonable we could simply clone it (see https://github.com/chinedufn/percy/compare/master...dbrgn:patch-ownership), but I'm not sure if that's worth the improved API because this way the element always gets cloned, while it might not need to be cloned if the user doesn't need the reference to the element anymore after patching.

Another option would be to accept the element by value, to create an Rc from it, and to pass that Rc to the `find_nodes` function. Then in the end we could return the original element by value. But every Rc causes an allocation and I'm not even sure how to unwrap the Rc to recover the original value :slightly_smiling_face: So I think it might be best to stick to the current approach unless someone figures out lifetime annotations to be able to stick to references only.

TLDR: The user can clone the element before passing it to the `patch` function.

**Comment by chinedufn** [comment]
SWEET thanks for looking into this!

Ah so it was a lifetime issue - I vaguely remember haha ..

Sweet this sounds great for now 👍🏿
